File: apps/web/src/lib/social/recurrence.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	expandInstances,
	nextInstanceAfter,
	describeRecurrence,
} from './recurrence';
import type { Event } from '../types';

// Build a partial Event with just enough fields for recurrence math.
// Casts the rest to `unknown as Event` — the recurrence helpers only
// touch `starts_at`, `recurrence_freq`, `recurrence_byday`,
// `recurrence_until`, and `recurrence_count`.
function ev(partial: Partial<Event>): Event {
	return partial as unknown as Event;
}

test('expandInstances — non-recurring event returns its single instance when in window', () => {
	const e = ev({ starts_at: '2026-04-10T08:00:00Z' });
	const inWindow = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-04-30T23:59:59Z'),
	);
	assert.equal(inWindow.length, 1);

	const outOfWindow = expandInstances(
		e,
		new Date('2026-05-01T00:00:00Z'),
		new Date('2026-05-31T23:59:59Z'),
	);
	assert.equal(outOfWindow.length, 0);
});

test('expandInstances — weekly event produces multiple instances in a month', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-04-30T23:59:59Z'),
	);
	// Apr 1 (Wed), Apr 8, Apr 15, Apr 22, Apr 29 → 5 instances. The Apr 1
	// instance used to slip on UTC hosts because the loop compared a
	// midnight cursor against starts_at's 08:00 time-of-day — the fix
	// separates calendar-day and stamped-time comparisons so the first-
	// week instance is no longer silently dropped.
	assert.equal(out.length, 5);
});

test('expandInstances — weekly emits the starts_at week even when start has a non-midnight time-of-day', () => {
	// Regression: this used to drop the Tuesday instance because the
	// midnight cursor for the start day was < starts_at (Tuesday 19:00),
	// so the entire first week's matching day was filtered out before
	// time-of-day was stamped.
	const e = ev({
		starts_at: '2026-04-07T19:00:00Z', // Tue
		recurrence_freq: 'weekly',
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-04-30T23:59:59Z'),
	);
	const firstDay = out[0]?.toISOString().slice(0, 10);
	assert.equal(firstDay, '2026-04-07', `first instance should be Apr 7, got ${firstDay}`);
});

test('expandInstances — biweekly produces fewer instances than weekly in the same window', () => {
	const start = '2026-04-01T08:00:00Z';
	const weekly = ev({ starts_at: start, recurrence_freq: 'weekly' });
	const biweekly = ev({ starts_at: start, recurrence_freq: 'biweekly' });
	const from = new Date('2026-04-01T00:00:00Z');
	const to = new Date('2026-04-30T23:59:59Z');
	const w = expandInstances(weekly, from, to);
	const b = expandInstances(biweekly, from, to);
	assert.ok(b.length < w.length, `biweekly ${b.length} should be fewer than weekly ${w.length}`);
});

test('expandInstances — biweekly with a weekend-crossing byday anchors on Monday', () => {
	// Twin-parity contract with apps/mobile_android/lib/recurrence.dart: the week
	// is anchored on Monday, so a biweekly [SA, SU] event starting on a Saturday
	// fires Sat+Sun on the start week, skips a week, then Sat+Sun again. A Sunday
	// anchor (the old Dart behaviour) split the Sunday into the next week and
	// produced a different set of dates. CI runs in UTC; the start time is given
	// in Z so getDay() reads Saturday.
	const e = ev({
		starts_at: '2026-05-02T09:00:00Z', // Sat
		recurrence_freq: 'biweekly',
		recurrence_byday: ['SA', 'SU'],
	});
	const out = expandInstances(
		e,
		new Date('2026-05-01T00:00:00Z'),
		new Date('2026-05-31T23:59:59Z'),
	);
	const dates = out.map((d) => d.toISOString().slice(0, 10));
	assert.deepEqual(dates, [
		'2026-05-02',
		'2026-05-03',
		'2026-05-16',
		'2026-05-17',
		'2026-05-30',
		'2026-05-31',
	]);
});

test('expandInstances — recurrence_count caps the number of instances', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
		recurrence_count: 3,
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-12-31T23:59:59Z'),
	);
	assert.equal(out.length, 3);
});

test('expandInstances — recurrence_until stops the expansion', () => {
	const until = '2026-04-22T08:00:00Z';
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
		recurrence_until: until,
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-12-31T23:59:59Z'),
	);
	assert.ok(out.length > 0);
	const untilMs = new Date(until).getTime();
	for (const d of out) {
		assert.ok(d.getTime() <= untilMs, `instance ${d.toISOString()} after until`);
	}
});

test('expandInstances — monthly produces one instance per month in window', () => {
	const e = ev({
		starts_at: '2026-04-05T10:00:00Z',
		recurrence_freq: 'monthly',
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-06-30T23:59:59Z'),
	);
	assert.equal(out.length, 3);
});

test('expandInstances — day-31 monthly clamps to month-end (no overflow)', () => {
	// Jan-31 monthly must land on Feb-28 (2026 is not a leap year), Mar-31,
	// Apr-30 — never roll into the following month. A naive setMonth(+1) on a
	// Jan-31 cursor overflows to March 3. Local-time start (no Z) so the
	// day-of-month is unambiguous in the UTC CI runner.
	const e = ev({
		starts_at: '2026-01-31T09:00:00',
		recurrence_freq: 'monthly',
	});
	const out = expandInstances(
		e,
		new Date('2026-01-01T00:00:00'),
		new Date('2026-04-30T23:59:59'),
	);
	const dates = out.map((d) => `${d.getFullYear()}-${d.getMonth() + 1}-${d.getDate()}`);
	assert.deepEqual(dates, ['2026-1-31', '2026-2-28', '2026-3-31', '2026-4-30']);
	// Each instance re-anchors on the original day (March is back to 31), so a
	// February clamp doesn't permanently shrink the day-of-month.
	for (const d of out) {
		assert.equal(d.getHours(), 9);
		assert.equal(d.getMinutes(), 0);
	}
});

test('expandInstances — day-31 monthly hits Feb-29 in a leap year', () => {
	const e = ev({
		starts_at: '2024-01-31T08:00:00',
		recurrence_freq: 'monthly',
	});
	const out = expandInstances(
		e,
		new Date('2024-02-01T00:00:00'),
		new Date('2024-02-29T23:59:59'),
	);
	assert.equal(out.length, 1);
	assert.equal(out[0].getMonth(), 1); // February
	assert.equal(out[0].getDate(), 29);
});

test('expandInstances — instances do not precede the original starts_at', () => {
	const startsAt = '2026-04-08T09:00:00Z';
	const e = ev({ starts_at: startsAt, recurrence_freq: 'weekly' });
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-04-30T23:59:59Z'),
	);
	const startMs = new Date(startsAt).getTime();
	for (const d of out) {
		assert.ok(d.getTime() >= startMs, `instance ${d.toISOString()} precedes starts_at`);
	}
});

test('expandInstances — monthly with recurrence_count caps instances', () => {
	const e = ev({
		starts_at: '2026-01-01T09:00:00Z',
		recurrence_freq: 'monthly',
		recurrence_count: 4,
	});
	const out = expandInstances(
		e,
		new Date('2026-01-01T00:00:00Z'),
		new Date('2026-12-31T23:59:59Z'),
	);
	assert.equal(out.length, 4);
});

test('nextInstanceAfter — returns the next instance after a cutoff', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
	});
	const next = nextInstanceAfter(e, new Date('2026-04-10T00:00:00Z'));
	assert.ok(next != null);
	// The first weekly instance after Apr 10 should be Apr 15 (Wed).
	assert.equal(next!.toISOString().slice(0, 10), '2026-04-15');
});

test('nextInstanceAfter — returns null past recurrence_until', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
		recurrence_until: '2026-04-30T23:59:59Z',
	});
	const next = nextInstanceAfter(e, new Date('2026-05-01T00:00:00Z'));
	assert.equal(next, null);
});

test('describeRecurrence — null freq is "One-off event"', () => {
	assert.equal(describeRecurrence(null, []), 'One-off event');
	assert.equal(describeRecurrence(null, null), 'One-off event');
});

test('describeRecurrence — monthly is "Repeats monthly"', () => {
	assert.equal(describeRecurrence('monthly', null), 'Repeats monthly');
	assert.equal(describeRecurrence('monthly', ['MO']), 'Repeats monthly');
});

test('describeRecurrence — weekly with byday lists days in ISO order', () => {
	// Even when input order is shuffled, output order is MO,TU,...,SU.
	assert.equal(describeRecurrence('weekly', ['WE', 'MO']), 'Every week · Mon, Wed');
});

test('describeRecurrence — biweekly without byday is bare', () => {
	assert.equal(describeRecurrence('biweekly', null), 'Every other week');
	assert.equal(describeRecurrence('biweekly', []), 'Every other week');
});

File: apps/mobile_android/test/recurrence_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/recurrence.dart';

void main() {
  group('expandInstances — weekly recurrence', () {
    test('each weekly instance hour/minute matches the local interpretation of startsAt', () {
      // An event stored in the DB as UTC (e.g. 10:30 UTC).
      // After the fix, each instance carries the *local* hour/minute of that UTC
      // timestamp. Before the fix, it carried the raw UTC hour, which is wrong
      // for any user whose offset != 0.
      final utcStart = DateTime.utc(2026, 4, 1, 10, 30, 0);
      final localStart = utcStart.toLocal();
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 4, 30);

      final instances = expandInstances(e, from, to);
      expect(instances, isNotEmpty,
          reason: 'at least one weekly instance should fall in April');
      for (final inst in instances) {
        // inst is a local (unzoned) DateTime — its .hour/.minute are local fields.
        expect(inst.hour, localStart.hour,
            reason: 'instance hour should be localStart.hour, not the raw UTC hour');
        expect(inst.minute, localStart.minute);
        expect(inst.second, localStart.second);
      }
    });

    test('instances do not precede the original startsAt', () {
      final utcStart = DateTime.utc(2026, 4, 8, 9, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 4, 30);

      final instances = expandInstances(e, from, to);
      expect(instances, isNotEmpty);
      for (final inst in instances) {
        expect(inst.isBefore(e.startsAt), isFalse,
            reason: 'no instance should precede the event start');
      }
    });

    test('biweekly event produces fewer instances than weekly in the same window', () {
      final utcStart = DateTime.utc(2026, 4, 1, 8, 0, 0);
      final weekly = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
      );
      final biweekly = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.biweekly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 4, 30);

      final wInstances = expandInstances(weekly, from, to);
      final bInstances = expandInstances(biweekly, from, to);
      // Biweekly should produce at most half as many instances.
      expect(bInstances.length, lessThan(wInstances.length));
    });

    test('biweekly with a weekend-crossing byday matches the web Monday anchor', () {
      // Twin-parity regression: Dart used to anchor the week on Sunday while
      // web anchors on Monday, so a biweekly [SA, SU] event disagreed on which
      // alternating weeks fire. Local (unzoned) start so the byday is the same
      // wall-clock Saturday in every test timezone. Expected dates are the
      // canonical web output (Monday anchor): May 2/3, 16/17, 30/31 2026.
      final e = EventRecurrence(
        startsAt: DateTime(2026, 5, 2, 9, 0), // Sat 09:00 local
        freq: RecurrenceFreq.biweekly,
        byday: const [Weekday.sa, Weekday.su],
      );

      final instances = expandInstances(
        e,
        DateTime(2026, 5, 1),
        DateTime(2026, 5, 31, 23, 59, 59),
      );

      final dates = instances.map((d) => '${d.year}-${d.month}-${d.day}').toList();
      expect(dates, ['2026-5-2', '2026-5-3', '2026-5-16', '2026-5-17', '2026-5-30', '2026-5-31']);
    });

    test('recurrence_count cap limits the number of instances', () {
      final utcStart = DateTime.utc(2026, 4, 1, 8, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
        count: 2,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 12, 31);

      final instances = expandInstances(e, from, to);
      expect(instances, hasLength(2));
    });

    test('until date stops expansion — no instance falls after until', () {
      final utcStart = DateTime.utc(2026, 4, 1, 8, 0, 0);
      final until = DateTime.utc(2026, 4, 22);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
        until: until,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 12, 31);

      final instances = expandInstances(e, from, to);
      expect(instances, isNotEmpty);
      for (final inst in instances) {
        expect(inst.isAfter(until), isFalse,
            reason: 'no instance should fall after the until date');
      }
    });

    test('non-recurring event returns the single instance if in window', () {
      final utcStart = DateTime.utc(2026, 4, 10, 8, 0, 0);
      final e = EventRecurrence(startsAt: utcStart);

      expect(
        expandInstances(e, DateTime.utc(2026, 4, 1), DateTime.utc(2026, 4, 30)),
        hasLength(1),
      );
      expect(
        expandInstances(e, DateTime.utc(2026, 5, 1), DateTime.utc(2026, 5, 31)),
        isEmpty,
      );
    });
  });

  group('expandInstances — monthly recurrence', () {
    test('monthly event produces one instance per month in window', () {
      final utcStart = DateTime.utc(2026, 4, 5, 10, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.monthly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 6, 30);

      final instances = expandInstances(e, from, to);
      expect(instances, hasLength(3)); // Apr 5, May 5, Jun 5
    });

    test('day-31 monthly clamps to the last day of shorter months (no overflow)', () {
      // Jan-31 monthly must land on Feb-28 (2026 is not a leap year), Mar-31,
      // Apr-30 — never overflow into the following month. A naive
      // DateTime(y, 2, 31) rolls into March 3.
      final start = DateTime(2026, 1, 31, 9, 0);
      final e = EventRecurrence(
        startsAt: start,
        freq: RecurrenceFreq.monthly,
      );

      final instances = expandInstances(
        e,
        DateTime(2026, 1, 1),
        DateTime(2026, 4, 30, 23, 59, 59),
      );

      final dates = instances.map((d) => '${d.year}-${d.month}-${d.day}').toList();
      expect(dates, ['2026-1-31', '2026-2-28', '2026-3-31', '2026-4-30']);
      // The clamp must not permanently shrink the day-of-month: March is
      // back to 31, proving each instance re-anchors on the original day.
      for (final inst in instances) {
        expect(inst.hour, 9);
        expect(inst.minute, 0);
      }
    });

    test('day-31 monthly hits Feb-29 in a leap year', () {
      final start = DateTime(2024, 1, 31, 8, 0);
      final e = EventRecurrence(
        startsAt: start,
        freq: RecurrenceFreq.monthly,
      );

      final instances = expandInstances(
        e,
        DateTime(2024, 2, 1),
        DateTime(2024, 2, 29, 23, 59, 59),
      );

      expect(instances, hasLength(1));
      expect(instances.first.month, 2);
      expect(instances.first.day, 29);
    });

    test('monthly recurrence_count cap limits instances', () {
      final utcStart = DateTime.utc(2026, 1, 1, 9, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.monthly,
        count: 3,
      );

      final from = DateTime.utc(2026, 1, 1);
      final to = DateTime.utc(2026, 12, 31);

      final instances = expandInstances(e, from, to);
      expect(instances, hasLength(3));
    });
  });

  group('expandInstances — timezoned (viewer-independent instance_start)', () {
    test('timezoned weekly anchors instance_start to UTC wall-clock', () {
      // The capacity key + race-arm key for a recurring instance must be the
      // same instant for every spectator. With a timezone present the
      // expansion reads + stamps the fields in UTC, so the produced instants
      // are fixed in absolute time and don't depend on the device's zone.
      // Asserting the exact UTC ISO instants pins that under any test TZ.
      final e = EventRecurrence(
        startsAt: DateTime.utc(2026, 4, 7, 13, 0, 0), // Tue 13:00 UTC
        freq: RecurrenceFreq.weekly,
        timezone: 'America/New_York',
      );
      final out = expandInstances(
        e,
        DateTime.utc(2026, 4, 1),
        DateTime.utc(2026, 4, 30, 23, 59, 59),
      );
      final iso = out.map((d) => d.toUtc().toIso8601String()).toList();
      expect(iso, [
        '2026-04-07T13:00:00.000Z',
        '2026-04-14T13:00:00.000Z',
        '2026-04-21T13:00:00.000Z',
        '2026-04-28T13:00:00.000Z',
      ]);
    });

    test('two cross-TZ viewers compute the SAME instance_start', () {
      // The bug: the .toLocal() + local-field stamp built instants in the
      // VIEWER's zone, so spectators in different zones RSVP'd against
      // different capacity keys. For a timezoned event the expansion reads no
      // ambient zone other than UTC, invariant across viewers. Re-running the
      // same input twice (standing in for two viewers) yields identical
      // instants; the absolute-time assertion guards the cross-device case
      // (the suite is also run under multiple TZ values).
      final e = EventRecurrence(
        startsAt: DateTime.utc(2026, 5, 2, 23, 30, 0), // Sat 23:30 UTC — a
        freq: RecurrenceFreq.biweekly, // viewer-local stamp would roll the day.
        byday: const [Weekday.sa],
        timezone: 'Europe/London',
      );
      final from = DateTime.utc(2026, 5, 1);
      final to = DateTime.utc(2026, 5, 31, 23, 59, 59);
      final viewerA = expandInstances(e, from, to).map((d) => d.toUtc().toIso8601String()).toList();
      final viewerB = expandInstances(e, from, to).map((d) => d.toUtc().toIso8601String()).toList();
      expect(viewerA, viewerB);
      expect(viewerA, [
        '2026-05-02T23:30:00.000Z',
        '2026-05-16T23:30:00.000Z',
        '2026-05-30T23:30:00.000Z',
      ]);
    });

    test('timezoned monthly anchors day-of-month + time in UTC', () {
      final e = EventRecurrence(
        startsAt: DateTime.utc(2026, 1, 31, 22, 0, 0),
        freq: RecurrenceFreq.monthly,
        timezone: 'Australia/Sydney',
      );
      final out = expandInstances(
        e,
        DateTime.utc(2026, 1, 1),
        DateTime.utc(2026, 4, 30, 23, 59, 59),
      );
      final iso = out.map((d) => d.toUtc().toIso8601String()).toList();
      // Jan-31, Feb-28 (clamp, 2026 non-leap), Mar-31, Apr-30 — all 22:00 UTC.
      expect(iso, [
        '2026-01-31T22:00:00.000Z',
        '2026-02-28T22:00:00.000Z',
        '2026-03-31T22:00:00.000Z',
        '2026-04-30T22:00:00.000Z',
      ]);
    });

    test('legacy event with no timezone keeps viewer-local behaviour', () {
      // A row predating 20270111_001 has no timezone; expansion falls back to
      // the original local-zone stamping so already-placed RSVPs don't shift.
      final e = EventRecurrence(
        startsAt: DateTime.utc(2026, 4, 7, 13, 0, 0),
        freq: RecurrenceFreq.weekly,
      );
      final out = expandInstances(
        e,
        DateTime.utc(2026, 4, 1),
        DateTime.utc(2026, 4, 30, 23, 59, 59),
      );
      expect(out, hasLength(4));
      final localStart = e.startsAt.toLocal();
      for (final inst in out) {
        expect(inst.hour, localStart.hour);
        expect(inst.isUtc, isFalse);
      }
    });
  });
}
